pkg/resources/provisioning.cattle.io/v1/cluster/validator.go

```diff
@@ -194,11 +215,100 @@ func (p *provisioningAdmitter) validateDataDirectories(request *admission.Reques
 	if response := p.validateSystemAgentDataDirectory(oldCluster, newCluster); !response.Allowed {
 		return response
 	}
-	if oldCluster.Spec.RKEConfig.DataDirectories.Provisioning != newCluster.Spec.RKEConfig.DataDirectories.Provisioning {
+	if oldCluster.Spec.RKEConfig.DataDirectories.K8sDistro != distro {
+		return admission.ResponseBadRequest("Distro data directory cannot be changed after cluster creation")
+	}
+	if oldCluster.Spec.RKEConfig.DataDirectories.Provisioning != provisioning {
 		return admission.ResponseBadRequest("Provisioning data directory cannot be changed after cluster creation")
 	}
-	if oldCluster.Spec.RKEConfig.DataDirectories.K8sDistro != newCluster.Spec.RKEConfig.DataDirectories.K8sDistro {
-		return admission.ResponseBadRequest("Distro data directory cannot be changed after cluster creation")
+
+	return admission.ResponseAllowed()
+}
+
+// validateDataDirectoryFormat ensures that no data directory contains a relative path, environment variables,
+// shell expressions, or references to the current or parent directory via use of "./" and "../" respectively.
+// dir is the path of the data directory, and name corresponds to a print friendly name for this data directory.
+func validateDataDirectoryFormat(dir, name string) *admissionv1.AdmissionResponse {
+	if dir == "" {
+		return admission.ResponseAllowed()
+	}
+	if !strings.HasPrefix(dir, "/") {
+		return admission.ResponseBadRequest(
+			fmt.Sprintf("%s data directory must be an absolute path", name))
+	}
+	if strings.HasSuffix(dir, "/") {
+		return admission.ResponseBadRequest(
+			fmt.Sprintf("%s data directory cannot end with \"/\"", name))
+	}
+
+	// Although this approach is somewhat naive, it ensures that any attempts to use environment variables as path names
+	// are rejected
+	if strings.Contains(dir, "$") || strings.Contains(dir, "${") {
+		return admission.ResponseBadRequest(
+			fmt.Sprintf("%s data directory cannot contain environment variables", name))
+	}
+
+	// As with the above, although valid directories can contain these characters under Linux, they are prohibited for
+	// simplicity's sake
+	if strings.Contains(dir, "`") || strings.Contains(dir, "$(") {
+		return admission.ResponseBadRequest(
+			fmt.Sprintf("%s data directory cannot contain shell expressions", name))
+	}
+
+	// Hidden directories (e.g. "/var/lib/rancher/.rke2") should be allowed, however relative paths should not be
+	if strings.Contains(dir, "/./") || strings.HasSuffix(dir, "/.") {
+		return admission.ResponseBadRequest(
+			fmt.Sprintf("%s data directory cannot contain reference to current directory \"(./)\"", name))
+	}
+
+	// As with the above, a directory like "/var/lib/rancher/..-rke2" is technically a valid folder name
+	if strings.Contains(dir, "/../") || strings.HasSuffix(dir, "/..") {
+		return admission.ResponseBadRequest(
+			fmt.Sprintf("%s data directory cannot contain reference to parent directory \"(../)\"", name))
+	}
```

Comment:
Can we just use https://pkg.go.dev/path/filepath#IsAbs and https://pkg.go.dev/path/filepath#Clean? Make sure it's absolute and is unchanged by cleaning.  Could probably just use strings.ContainsAny for the shell char checks too.

Example: https://go.dev/play/p/6bY3TQ89idh

Reply:
Updated. I kept the 3 checks as 3 separate if statements for error handling purposes, moving `filepath.Clean` to the last since it's the most esoteric.